File: src/main/kotlin/com/xs0/gqlktx/schema/intro/GqlIntroSchema.java
package com.xs0.gqlktx.schema.intro;

import com.xs0.gqlktx.ann.GraphQLObject;
import com.xs0.gqlktx.schema.Schema;
import com.xs0.gqlktx.types.gql.GBaseType;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.xs0.gqlktx.schema.intro.GqlIntroDirectiveLocation.FIELD;
import static com.xs0.gqlktx.schema.intro.GqlIntroDirectiveLocation.FRAGMENT_SPREAD;
import static com.xs0.gqlktx.schema.intro.GqlIntroDirectiveLocation.INLINE_FRAGMENT;
import static java.util.Collections.singletonList;

@GraphQLObject("__Schema")
public class GqlIntroSchema {
    private final Schema<?> schema;
    private final List<GqlIntroDirective> directives;

    public GqlIntroSchema(Schema<?> schema) {
        this.schema = schema;
        this.directives = buildDirectives();
    }

    private List<GqlIntroDirective> buildDirectives() {
        ArrayList<GqlIntroDirective> res = new ArrayList<>();

        GqlIntroType boolNotNull = schema.getGQLBaseType("Boolean").notNull().introspector();

        res.add(
            new GqlIntroDirective(
                "if",
                "Directs the executor to include this field or fragment only when the `if` argument is true.",
                Arrays.asList(FIELD, FRAGMENT_SPREAD, INLINE_FRAGMENT),
                singletonList(new GqlIntroInputValue("if", "Included when true.", boolNotNull, null))
            )
        );

        res.add(
            new GqlIntroDirective(
                "skip",
                "Directs the executor to skip this field or fragment when the `if` argument is true.",
                Arrays.asList(FIELD, FRAGMENT_SPREAD, INLINE_FRAGMENT),
                singletonList(new GqlIntroInputValue("if", "Skipped when true.", boolNotNull, null))
            )
        );

        return res;
    }

    public List<GqlIntroType> getTypes() {
        ArrayList<GqlIntroType> res = new ArrayList<>();

        for (GBaseType type : schema.getAllBaseTypes())
            res.add(type.introspector());

        return res;
    }

    public GqlIntroType getQueryType() {
        return schema.getJavaType(schema.getQueryRoot().getType()).getGqlType().introspector();
    }

    public GqlIntroType getMutationType() {
        if (schema.getMutationRoot() == null)
            return null;

        return schema.getJavaType(schema.getMutationRoot().getType()).getGqlType().introspector();
    }

    public GqlIntroType getSubscriptionType() {
        return null; // TODO
    }

    public List<GqlIntroDirective> getDirectives() {
        return directives;
    }

    public GqlIntroSchema self() {
        return this; // used for simple implementation
    }

    public GqlIntroType type(String name) {
        GBaseType baseType = schema.getGQLBaseType(name);
        return baseType != null ? baseType.introspector() : null;
    }
}

File: src/main/kotlin/com/xs0/gqlktx/schema/intro/GqlIntroSchema.kt
package com.xs0.gqlktx.schema.intro

import com.xs0.gqlktx.GqlObject
import com.xs0.gqlktx.schema.Schema
import com.xs0.gqlktx.schema.builder.TypeKind
import com.xs0.gqlktx.types.gql.GBaseType

import java.util.ArrayList
import java.util.Arrays

import com.xs0.gqlktx.schema.intro.GqlIntroDirectiveLocation.FIELD
import com.xs0.gqlktx.schema.intro.GqlIntroDirectiveLocation.FRAGMENT_SPREAD
import com.xs0.gqlktx.schema.intro.GqlIntroDirectiveLocation.INLINE_FRAGMENT
import com.xs0.gqlktx.types.kotlin.GJavaNotNullType
import java.util.Collections.singletonList

@GqlObject("__Schema")
class GqlIntroSchema(private val schema: Schema<*, *>) {
    val directives: List<GqlIntroDirective> = buildDirectives()

    private fun buildDirectives(): List<GqlIntroDirective> {
        val res = ArrayList<GqlIntroDirective>()

        val boolNotNull = schema.getGQLBaseType("Boolean").notNull().introspector

        res.add(
                GqlIntroDirective(
                        "if",
                        "Directs the executor to include this field or fragment only when the `if` argument is true.",
                        listOf(FIELD, FRAGMENT_SPREAD, INLINE_FRAGMENT),
                        listOf(GqlIntroInputValue("if", "Included when true.", boolNotNull, null))
                )
        )

        res.add(
                GqlIntroDirective(
                        "skip",
                        "Directs the executor to skip this field or fragment when the `if` argument is true.",
                        listOf(FIELD, FRAGMENT_SPREAD, INLINE_FRAGMENT),
                        listOf(GqlIntroInputValue("if", "Skipped when true.", boolNotNull, null))
                )
        )

        return res
    }

    fun getTypes(kinds: Array<TypeKind>?): List<GqlIntroType> {
        val res = ArrayList<GqlIntroType>()
        val filter: (TypeKind)->Boolean
        if (kinds != null && kinds.isNotEmpty()) {
            filter = kinds.toSet()::contains
        } else {
            filter = { _ -> true }
        }

        for (type in schema.allBaseTypes)
            if (filter(type.kind))
                res.add(type.introspector)

        return res
    }

    val queryType: GqlIntroType
        get() {
            var type = schema.getJavaType(schema.queryRoot.type)
            if (type is GJavaNotNullType)
                type = type.innerType

            return type.gqlType.introspector
        }

    val mutationType: GqlIntroType?
        get() {
            if (schema.mutationRoot == null)
                return null

            var type = schema.getJavaType(schema.mutationRoot.type)
            if (type is GJavaNotNullType)
                type = type.innerType

            return type.gqlType.introspector
        }

    // TODO
    val subscriptionType: GqlIntroType?
        get() = null

    fun self(): GqlIntroSchema {
        return this // used for simple implementation
    }

    fun type(name: String): GqlIntroType {
        return schema.getGQLBaseType(name).introspector
    }
}
